### PR TITLE
PLANNER-1827: Add ConstraintProvider for Investment Example.

### PR DESCRIPTION
**Pull request by Christopher-Chianelli**

### CODE REVIEW

File: optaplanner-examples/src/main/java/org/optaplanner/examples/investment/optional/score/InvestmentConstraintProvider.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                                assetsDeviationGreaterThanMaximumPenalty(constraintFactory),
+                                regionQuantityGreaterThanMaximumPenalty(constraintFactory),
+                                sectorQuantityGreaterThanMaximumPenalty(constraintFactory),
+                                assetExpectedReturnReward(constraintFactory)
+        };
+    }
+
+    private Constraint assetsDeviationGreaterThanMaximumPenalty(ConstraintFactory cf) {
+        return cf.from(AssetClassAllocation.class)
+                 .join(AssetClassAllocation.class)
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/investment/optional/score/InvestmentConstraintProvider.java]: `.fromUniquePair(AssetClassAllocation.class)`?

File: optaplanner-examples/src/main/java/org/optaplanner/examples/investment/optional/score/InvestmentConstraintProvider.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import org.optaplanner.examples.investment.domain.Region;
+import org.optaplanner.examples.investment.domain.Sector;
+
+public class InvestmentConstraintProvider implements ConstraintProvider {
+
+    private static final String CONSTRAINT_PACKAGE = "org.optaplanner.examples.investment.solver";
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        // TODO Auto-generated method stub
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/investment/optional/score/InvestmentConstraintProvider.java]: Please remove.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/investment/optional/score/InvestmentConstraintProvider.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        return new Constraint[]{
+                                assetsDeviationGreaterThanMaximumPenalty(constraintFactory),
+                                regionQuantityGreaterThanMaximumPenalty(constraintFactory),
+                                sectorQuantityGreaterThanMaximumPenalty(constraintFactory),
+                                assetExpectedReturnReward(constraintFactory)
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/investment/optional/score/InvestmentConstraintProvider.java]: Wrong indentation.
Also, I'm pretty sure some of the lines in this file exceed the 120-char limit.

Please configure your IDE according to the accepted code style.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/investment/optional/score/InvestmentConstraintProvider.java (hunk 1 above)
Comment (Christopher-Chianelli) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/investment/optional/score/InvestmentConstraintProvider.java]: Nope, duplicates are required for FULL_ASSERT to pass (apparently `a - a > 0` for investments)

File: optaplanner-examples/src/main/java/org/optaplanner/examples/investment/optional/score/InvestmentConstraintProvider.java (hunk 3 above)
Comment (Christopher-Chianelli) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/investment/optional/score/InvestmentConstraintProvider.java]: Again, this is KIE Java Conventions for Eclipse; will write a Dockerfile to run the IDEA formatter on a directory.

**Comment by Christopher-Chianelli** [comment]
It passes a 5 minute test on `FULL_ASSERT` (it auto-terminated after 5 due to solver config)

File: optaplanner-examples/src/main/java/org/optaplanner/examples/investment/optional/score/InvestmentConstraintProvider.java (hunk 1 above)
Comment (triceo) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/investment/optional/score/InvestmentConstraintProvider.java]: Would be nice to add a comment here then, explaining the situation.
I can already see some poor well-meaning soul "fixing" this unknowingly. :-)

File: optaplanner-examples/src/main/java/org/optaplanner/examples/investment/optional/score/InvestmentConstraintProvider.java (hunk 3 above)
Comment (Christopher-Chianelli) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/investment/optional/score/InvestmentConstraintProvider.java]: Done, configured a git hook to auto run IDEA formatter on changed files before commit in all KIE repo's.

**Pull request merged**